lib/collection/src/update_handler.rs

```diff
@@ -221,24 +230,55 @@ impl UpdateHandler {
     pub(crate) fn launch_optimization<F>(
         optimizers: Arc<Vec<Arc<Optimizer>>>,
         optimizers_log: Arc<Mutex<TrackerLog>>,
+        optimizer_cpu_budget: &CpuBudget,
         segments: LockedSegmentHolder,
         callback: F,
+        limit: Option<usize>,
     ) -> Vec<StoppableTaskHandle<bool>>
     where
-        F: FnOnce(bool),
-        F: Send + 'static,
-        F: Clone,
+        F: FnOnce(bool) + Send + Clone + 'static,
     {
         let mut scheduled_segment_ids: HashSet<_> = Default::default();
         let mut handles = vec![];
-        for optimizer in optimizers.iter() {
+        'outer: for optimizer in optimizers.iter() {
             loop {
+                // Return early if we reached the optimization job limit
+                if limit.map(|extra| handles.len() >= extra).unwrap_or(false) {
+                    log::trace!("Reached optimization job limit, postponing other optimizations");
+                    break 'outer;
+                }
+
                 let nonoptimal_segment_ids =
                     optimizer.check_condition(segments.clone(), &scheduled_segment_ids);
                 if nonoptimal_segment_ids.is_empty() {
                     break;
                 }
 
+                // Determine how many CPUs we prefer for optimization task, acquire permit for it
+                let max_indexing_threads = optimizer.hnsw_config().max_indexing_threads;
```

Comment:
What if `max_indexing_threads > optimizer_cpu_budget`?

Reply:
I think this could become important to explicitly handle once we have better logic in `try_acquire`

Reply:
This is fine in the latest implementation. The desired number of CPUs is internally limited to the total CPU budget.

For example:
- total CPU budget: 8
- desired: usize::MAX
- permit aquired: 4..=8